### PR TITLE
Generalize perturb and wrapped objective for more arguments

### PR DESCRIPTION
**Pull request by f0uriest**
- As we add more "stuff" (eg anisotropy, kinetic profiles etc) it's a pain to have to remember to also add args for those things in `perturb` and make sure each objective knows about them.
- This refactors the `perturb` function to take a dictionary of arrays rather than individual args for each thing, and makes sure the wrapped objective (and other objectives) know about all possible arguments.

### CODE REVIEW

File: desc/objectives/_wrappers.py [hunk 1]
```diff
@@ -89,7 +89,29 @@ def build(self, eq, use_jit=None, verbose=1):
...
+            self._args.remove("R_lmn")
+            if "Rb_lmn" not in self._args:
+                self._args.append("Rb_lmn")
+        if "Z_lmn" in self._args:
+            self._args.remove("Z_lmn")
+            if "Zb_lmn" not in self._args:
+                self._args.append("Zb_lmn")
+        # and remove pressure if we're solving a vacuum equilibrium
         if isinstance(self._eq_objective.objectives[0], CurrentDensity):
             self._args.remove("p_l")
```
Comment (ddudt) [inline desc/objectives/_wrappers.py:116]: Are these lines still necessary? If we are solving a vacuum equilibrium I think the list of args will already exclude pressure.

File: desc/perturbations.py [hunk 2]
```diff
@@ -58,13 +58,13 @@ def get_deltas(things1, things2):
                 if hasattr(t1, "change_resolution") and hasattr(t2, "basis"):
                     t1.change_resolution(t2.basis.L)
                 if not np.allclose(t2.params, t1.params):
-                    deltas["d" + key[0]] = t2.params - t1.params
+                    deltas[key[0] + "_l"] = t2.params - t1.params
```
Comment (ddudt) [inline desc/perturbations.py:61]: This works but might not generalize well (if we add another profile that starts with the letter "i", "p", or "c").  I'm not sure what a better method would be, but we should at least make a comment in the code like `# FIXME: generalize`

**Review by ddudt — CHANGES_REQUESTED** [review]

File: desc/perturbations.py (hunk 2 above)
Comment (f0uriest) [reply desc/perturbations.py:62]: added a note about this. We should try to come up with a more general way to map between profile names and the names of their parameters

File: desc/objectives/_wrappers.py (hunk 1 above)
Comment (f0uriest) [reply desc/objectives/_wrappers.py]: Yeah I think we can get rid of it.

**Review by dpanici — DISMISSED** [review]
QH optimization test is failing though?

**Comment by f0uriest** [comment]
Looks like its the 1d optimization failing now as well, stalling out the same as the others. Still looking into it.

File: desc/objectives/_wrappers.py [hunk 3]
```diff
@@ -89,18 +89,43 @@ def build(self, eq, use_jit=None, verbose=1):
...
+
+        # this is everything taken by either objective
+        self._full_args = (
+            self._eq_objective.args + self._objective.args + ["Rb_lmn", "Zb_lmn"]
+        )
+        self._full_args = [arg for arg in arg_order if arg in self._full_args]
+        # remove constraints that aren't necessary
+        self._constraints = tuple(
+            [con for con in self._constraints if con.args[0] in self._eq_objective.args]
+        )
```
[the remark this answers is not in the document]
Comment (f0uriest) [reply desc/objectives/_wrappers.py:101]: This is removing constraints that are targeting things the objective doesn't use so shouldn't be present in `x` anyways.

[the remark this answers is not in the document]
Comment (ddudt) [reply desc/objectives/_wrappers.py:101]: This is only keeping constraints that are in the equilibrium objective (force balance, etc.), and removing constraints only in the optimization objective (QS, etc.). I would think we want it to be the other way around:
`[con for con in self._constraints if con.args[0] in self._objective.args]`

[the remark this answers is not in the document]
Comment (ddudt) [reply desc/objectives/_wrappers.py:101]: Is there any harm to keeping all the constraints?

[the remark this answers is not in the document]
Comment (f0uriest) [reply desc/objectives/_wrappers.py:101]: Yes, these constraints are only used in the fixed boundary subproblem, which doesn't know anything about the outer objective

[the remark this answers is not in the document]
Comment (ddudt) [reply desc/objectives/_wrappers.py:101]: It looks like they get used in `WrappedEquilibriumObjective.jac`, so I think we need to make sure that `self._unfixed_idx` and the other outputs of `factorize_linear_constraints` correspond to the full optimization state vector. For example, these lines look like they could be problematic if `xg` has arguments that the equilibrium sub-problem is unaware of: 
```
Gx = self._objective.jac(xg)
Gx = align_jacobian(Gx, self._objective, self._eq_objective)
Gx_reduced = Gx[:, self._unfixed_idx] @ self._Z
```
(`align_jacobian` handles the differences between `xf` and `xg`, but I think `_unfixed_idx` is only based on the equilibrium sub-problem)

[the remark this answers is not in the document]
Comment (dpanici) [reply desc/objectives/_wrappers.py:101]: an example of things the eq objective does not use would be if someone threw a FixPressure constraint on an equilibrium that we are minimizing CurrentDensity in, where we do not need the pressure for the Equilibrium obejctive right? is this just to catch erroneous constraints like that?

**Review by dpanici — APPROVED** [review]

File: desc/objectives/_wrappers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (f0uriest) [reply desc/objectives/_wrappers.py:100]: @dpanici yes exactly. And @ddudt I fixed the issue you noticed, `self._unfixed_idx` is now built to broadcast against `self._full_args`, (which is what `Fx` and `Gx` are) not just `self._eq_objective.args` (which is usually the same as full_args but maybe not in all cases)

**Review by dpanici — APPROVED** [review]

**Review by ddudt — APPROVED** [review]

**Pull request merged**